File: pubnub-java/src/main/java/com/pubnub/api/managers/DuplicationManager.java
package com.pubnub.api.managers;

import com.pubnub.api.PNConfiguration;
import com.pubnub.api.models.server.SubscribeMessage;

import java.util.ArrayList;

public class DuplicationManager {

    private ArrayList<String> hashHistory;
    private PNConfiguration pnConfiguration;

    public DuplicationManager(PNConfiguration pnc) {
        this.hashHistory = new ArrayList<>();
        this.pnConfiguration = pnc;
    }

    private String getKey(SubscribeMessage message) {
        return message.getPublishMetaData().getPublishTimetoken().toString().concat("-").concat(Integer.toString(message.getPayload().hashCode()));
    }

    public boolean isDuplicate(SubscribeMessage message) {
        return hashHistory.contains(this.getKey(message));
    }

    public void addEntry(SubscribeMessage message) {
        if (this.hashHistory.size() >= pnConfiguration.getMaximumMessagesCacheSize()) {
            hashHistory.remove(0);
        }

        hashHistory.add(this.getKey(message));
    }

    public void clearHistory() {
        this.hashHistory.clear();
    }

}

File: core/src/main/kotlin/com/pubnub/api/managers/DuplicationManager.kt
package com.pubnub.api.managers

import com.pubnub.api.PNConfiguration
import com.pubnub.api.models.server.SubscribeMessage

internal class DuplicationManager(private val config: PNConfiguration) {

    private val hashHistory: ArrayList<String> = ArrayList()

    private fun getKey(message: SubscribeMessage) =
        with(message) {
            "${publishMetaData?.publishTimetoken}-${payload.hashCode()}"
        }

    @Synchronized
    fun isDuplicate(message: SubscribeMessage) = hashHistory.contains(getKey(message))

    @Synchronized
    fun addEntry(message: SubscribeMessage) {
        if (hashHistory.size >= config.maximumMessagesCacheSize) {
            hashHistory.removeAt(0)
        }
        hashHistory.add(getKey(message))
    }

    @Synchronized
    fun clearHistory() = hashHistory.clear()
}
